src/read.c

```diff
@@ -0,0 +1,143 @@
+      fb->mq1_header = mq1_header;
+      break;
+    }
+    case 768: {
+      mq1q mq1_quad;
+      parse_mq1_quad(header, &mq1_quad);
+      memcpy(fb->dac0, &mq1_quad.dac0, sizeof(dac_rx));
+      memcpy(fb->dac1, &mq1_quad.dac1, sizeof(dac_rx));
+      memcpy(fb->dac2, &mq1_quad.dac2, sizeof(dac_rx));
+      memcpy(fb->dac3, &mq1_quad.dac3, sizeof(dac_rx));
+      fill_MQ1_quad_fields(mq1_header, 0, mq1_quad);
+      fb->mq1_header = mq1_header;
+      break;
+    }
+    default: {
+      deallocate_MQ1_fields(*mq1_header);
+      free(header);
+      header = NULL;
+      free(mq1_header);
+      mq1_header = NULL;
+      fprintf(stderr, "headersize not 384 or 768\n");
+      return;
+    }
+  }
+  free(header);
+  header = NULL;
+}
+
+void read_frame(FILE *mib_ptr, long offset, framebuffer *fb)
+{
+  if (!mib_ptr || !fb) {
+    fprintf(stderr, "Missing input in read_frame\n");
+    if (!mib_ptr)
+      fprintf(stderr, "NO MIB_PTR\n");
+    if (!fb)
+      fprintf(stderr, "NO fb\n");
+    return;
+  }
+
+  int headersize = *(fb->mq1_header->header_bytes);
+
+  int bufsize = (fb->mq1_header->pixel_depth[1] - '0') * 10 +
+                (fb->mq1_header->pixel_depth[2] - '0');
+  bufsize = bufsize / 8;
+
+  int detx = (int) *(fb->mq1_header->det_x);
+  int dety = (int) *(fb->mq1_header->det_y);
+
+  // move mib_ptr to the correct place
+  fseek(mib_ptr, offset + headersize, SEEK_SET);
+
+  // write data into buffer
+  uint8_t *raw_data = malloc(bufsize * detx * dety);
+  if (!raw_data) {
+    fprintf(stderr, "malloc failed for raw_data in read_frame\n");
+    return;
+  }
+
+  // TO-DO: add checks for corruption
+  fread(raw_data, sizeof(char), bufsize * detx * dety, mib_ptr);
+
+  for (int i = 0; i < dety; i++) {
+    for (int j = 0; j < detx; j++) {
+      size_t index       = (i * detx + j) * bufsize;
+      uint8_t *raw_bytes = &raw_data[index];
+      uint64_t value     = 0;
+
+      switch (bufsize) {
+        case 1: {
+          value                         = raw_bytes[0];
+          ((uint8_t **) fb->rows)[i][j] = (uint8_t) value;
+          break;
+        }
+        case 2: {
+          value                          = (raw_bytes[0] << 8) | raw_bytes[1];
+          ((uint16_t **) fb->rows)[i][j] = (uint16_t) value;
+          break;
+        }
+        case 4: {
+          value = (raw_bytes[0] << 24) | (raw_bytes[1] << 16) |
+                  (raw_bytes[2] << 8) | raw_bytes[3];
+          ((uint32_t **) fb->rows)[i][j] = (uint32_t) value;
+          break;
+        }
+        case 8: {
+          value =
+            ((uint64_t) raw_bytes[0] << 56) | ((uint64_t) raw_bytes[1] << 48) |
+            ((uint64_t) raw_bytes[2] << 40) | ((uint64_t) raw_bytes[3] << 32) |
+            ((uint64_t) raw_bytes[4] << 24) | ((uint64_t) raw_bytes[5] << 16) |
+            ((uint64_t) raw_bytes[6] << 8) | ((uint64_t) raw_bytes[7]);
+          ((uint64_t **) fb->rows)[i][j] = value;
+          break;
```

Comment:
This approach of handling the endianness is nice in term of reading the big-endian data byte by byte and interpreting them using the CPU native format.

My only comment on this is to refactor the conversions to different `static` (should only be used in reading raw data) and `inline` functions with names such as `convert_uint8` or similar to communicate what it is actually happening here. It also facilitates future testing.